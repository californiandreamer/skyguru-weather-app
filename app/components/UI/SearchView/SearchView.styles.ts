import { transparent, white } from 'app/constants/colors'
import {
  gap,
  largeGap,
  littleGap,
  smallFontSize,
  smallGap,
} from 'app/constants/values'
import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  container: {
    padding: largeGap,
    paddingBottom: gap,
  },
  searchArea: {
    paddingTop: littleGap,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'center',
  },
  locationIcon: {
    paddingRight: smallGap,
  },
  input: {
    height: 32,
    paddingTop: 2,
    flex: 1,
    backgroundColor: transparent,
    fontSize: smallFontSize,
    fontFamily: 'Gilroy-Regular',
    color: white,
  },
  spinner: {
    paddingBottom: largeGap,
  },
  resultsArea: {
    paddingTop: gap,
    paddingLeft: gap + smallGap,
  },
  resultsItem: {
    paddingVertical: smallGap,
  },
  locationButtonArea: {
    padding: gap,
    justifyContent: 'center',
    alignItems: 'center',
  },
  locationButton: {
    height: 32,
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  locationButtonText: {
    fontSize: smallFontSize,
    fontFamily: 'Gilroy-Regular',
    color: white,
  },
})
