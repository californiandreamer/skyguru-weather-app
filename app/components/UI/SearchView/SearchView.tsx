import { PinIcon } from 'app/assets'
import StyledText from 'app/components/HOC/Text'
import {
  defaultIconSize,
  pressOpacity,
  resultsQuantity,
  smallFontSize,
} from 'app/constants/values'
import { useGeolocation } from 'app/hooks'
import { IGeolocationData } from 'app/models'
import { fetchCurrentWeatherRequest } from 'app/store/actions/currentWeather'
import { fetchFutureWeatherRequest } from 'app/store/actions/futureWeather'
import { hideSearch } from 'app/store/actions/search'
import React, { useEffect, useRef, useState } from 'react'
import { View, TextInput, TouchableOpacity, Text } from 'react-native'
import { useDispatch } from 'react-redux'

import styles from './SearchView.styles'
import cities from '../../../data/cities.json'

export type CitiesT = {
  country: string
  lat: string
  lng: string
  city: string
}

const SearchView: React.FC = () => {
  const [value, setValue] = useState<string>('')
  const [results, setResults] = useState<CitiesT[]>([])
  const [position] = useGeolocation()

  const dispatch = useDispatch()
  const inputRef = useRef<TextInput>(null)

  const focusInput = () => inputRef?.current?.focus()

  const filterCities = (text: string): void => {
    setValue(text)

    if (text.length > 2) {
      const filtered: CitiesT[] = cities.filter(
        (item: { city: string | string[] }) => item.city.includes(text)
      )
      const firstItems = filtered.slice(0, resultsQuantity)

      setResults(firstItems)
    } else {
      setResults([])
    }
  }

  const handleResultPress = (city: CitiesT) => {
    const cityPosition: IGeolocationData = {
      latitude: parseFloat(city.lat),
      longitude: parseFloat(city.lng),
    }

    dispatch(fetchCurrentWeatherRequest({ position: cityPosition }))
    dispatch(fetchFutureWeatherRequest({ position: cityPosition }))
    dispatch(hideSearch())
  }

  const handleOwnLocationPress = () => {
    dispatch(fetchCurrentWeatherRequest({ position }))
    dispatch(fetchFutureWeatherRequest({ position }))
    dispatch(hideSearch())
  }

  useEffect(() => {
    focusInput()
  }, [])

  const renderSearchArea = () => (
    <View style={styles.searchArea}>
      <View style={styles.locationIcon}>
        <PinIcon width={defaultIconSize} height={defaultIconSize} />
      </View>
      <TextInput
        style={styles.input}
        value={value}
        ref={inputRef}
        placeholder="Search city"
        accessibilityLabel="answer input"
        onChangeText={filterCities}
      />
    </View>
  )

  const renderResults = () => (
    <View style={styles.resultsArea} testID="results area">
      {results.map((item, index) => {
        console.log('item', item)

        return (
          <TouchableOpacity
            style={styles.resultsItem}
            activeOpacity={pressOpacity}
            key={index}
            accessibilityLabel="result item"
            onPress={() => handleResultPress(item)}
          >
            <StyledText fontSize={smallFontSize}>
              {item.city}, {item.country}
            </StyledText>
          </TouchableOpacity>
        )
      })}
    </View>
  )

  const renderLocationButton = () => (
    <View style={styles.locationButtonArea}>
      <TouchableOpacity
        style={styles.locationButton}
        activeOpacity={pressOpacity}
        accessibilityLabel="location button"
        onPress={handleOwnLocationPress}
      >
        <Text style={styles.locationButtonText}>Use my location</Text>
      </TouchableOpacity>
    </View>
  )

  return (
    <View style={styles.container}>
      {renderSearchArea()}
      {renderResults()}
      {renderLocationButton()}
    </View>
  )
}

export default SearchView
