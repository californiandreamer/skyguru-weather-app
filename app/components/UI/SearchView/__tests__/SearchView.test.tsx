import { render, fireEvent, cleanup } from '@testing-library/react-native'
import configureStore from 'app/store'
import React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'

import SearchView from '../SearchView'

jest.useRealTimers()

jest.mock('app/hooks/useGeolocation', () => ({
  useGeolocation: () => [{ latitude: '52.2167', longitude: '21.0333' }],
}))

describe('SearchView component', () => {
  const store = configureStore()
  const textToCheck = 'War'
  const textToEnter = 'Warsaw'
  const autocompleteResult = 'Warsaw, Poland'

  const component = (
    <Provider store={store}>
      <SearchView />
    </Provider>
  )

  afterEach(cleanup)

  it('returns right city', () => {
    const tree = render(component)
    const { getByPlaceholderText, getByText } = tree
    const input = getByPlaceholderText('Search city')
    expect(input).toBeTruthy()

    fireEvent.changeText(input, textToEnter)
    const result = getByText(autocompleteResult)
    expect(result).toBeTruthy()
  })

  it('returns correct quantity of results', () => {
    const tree = render(component)
    const { getByPlaceholderText, getAllByLabelText } = tree
    const input = getByPlaceholderText('Search city')
    expect(input).toBeTruthy()

    fireEvent.changeText(input, textToCheck)
    const result = getAllByLabelText('result item')

    expect(result.length).toBe(2)
  })

  it('does not show results if it is not enough letters', () => {
    const tree = render(component)
    const { getByPlaceholderText, getByTestId } = tree
    const input = getByPlaceholderText('Search city')
    expect(input).toBeTruthy()

    fireEvent.changeText(input, 'Wa')
    const result = getByTestId('results area')
    expect(result.props.children.length).toBe(0)
  })

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <SearchView />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
