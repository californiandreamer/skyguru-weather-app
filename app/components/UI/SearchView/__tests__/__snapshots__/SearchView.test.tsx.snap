// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchView component renders correctly 1`] = `
<View
  style={
    Object {
      "padding": 40,
      "paddingBottom": 20,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "paddingTop": 5,
      }
    }
  >
    <View
      style={
        Object {
          "paddingRight": 10,
        }
      }
    >
      <SvgMock
        height={20}
        width={20}
      />
    </View>
    <TextInput
      accessibilityLabel="answer input"
      onChangeText={[Function]}
      placeholder="Search city"
      style={
        Object {
          "backgroundColor": "transparent",
          "color": "#EEE",
          "flex": 1,
          "fontFamily": "Gilroy-Regular",
          "fontSize": 20,
          "height": 32,
          "paddingTop": 2,
        }
      }
      value=""
    />
  </View>
  <View
    style={
      Object {
        "paddingLeft": 30,
        "paddingTop": 20,
      }
    }
    testID="results area"
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <View
      accessibilityLabel="location button"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "height": 32,
          "justifyContent": "center",
          "opacity": 1,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#EEE",
            "fontFamily": "Gilroy-Regular",
            "fontSize": 20,
          }
        }
      >
        Use my location
      </Text>
    </View>
  </View>
</View>
`;
